### PR TITLE
feat(cosmos): add pagination to precompiles

### LINKED ISSUE #547: feat(cosmos): add pagination to Cosmos requests in precompiles
in precompiles, e.g. bank precompile, there are Cosmos requests that can take pagination as a request param:
e.g.
https://github.com/berachain/polaris/blob/26526ee6c28d5c0fdf00a26bf4b5acb8507c1de9/cosmos/precompile/bank/bank.go#L245

### PR DESCRIPTION
**Pull request by anukul**
related - #547

### CODE REVIEW

File: cosmos/lib/pagination.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+	for paginationKey, i := []byte{}, 0; len(paginationKey) > 0 || i == 0; paginationKey, i = getNextPaginationKey(response), i+1 {
+		response, err = responseFetcher(paginationKey)
+		if err != nil {
+			return nil, err
+		}
+		responses = append(responses, response)
+	}
+
+	return responses, nil
+}
```
Comment (anukul) [inline cosmos/lib/pagination.go:37]: We can add another function argument here to reduce the boilerplate of parsing results from responses (e.g. `response.GetSupply`), but it seems to reduce readability for very little gain

```go
// Paginate returns the aggregate result of a paginated cosmos query
func Paginate[T PageResponse, U interface{}](
	responseFetcher func(paginationKey []byte) (T, error),
	resultGetter func(response T) []U,
) ([]U, error) {
	var result []U
	var response T
	var err error

	for paginationKey, i := []byte{}, 0; len(paginationKey) > 0 || i == 0; paginationKey, i = getNextPaginationKey(response), i+1 {
		response, err = responseFetcher(paginationKey)
		if err != nil {
			return nil, err
		}
		result = append(result, resultGetter(response)...)
	}

	return result, nil
}
```

```go
supply, err := cosmlib.Paginate[*banktypes.QueryTotalSupplyResponse, types.Coin](
	func(paginationKey []byte) (*banktypes.QueryTotalSupplyResponse, error) {
		return c.querier.TotalSupply(ctx, &banktypes.QueryTotalSupplyRequest{
			Pagination: &query.PageRequest{Key: paginationKey},
		})
	},
	func(response *banktypes.QueryTotalSupplyResponse) []types.Coin {
		return response.GetSupply()
	},
)
```

File: cosmos/lib/pagination.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+	return pagination.NextKey
+}
+
+// Paginate returns the aggregate responses of a paginated cosmos query
+func Paginate[T PageResponse](
+	responseFetcher func(paginationKey []byte) (T, error),
+) (responses []T, err error) {
+	var response T
+
+	for paginationKey, i := []byte{}, 0; len(paginationKey) > 0 || i == 0; paginationKey, i = getNextPaginationKey(response), i+1 {
```
Comment (anukul) [inline cosmos/lib/pagination.go:28]: I'm not sure about the unbounded loop here (only termination condition being `paginationKey`, regardless of number of iterations)

File: cosmos/lib/pagination.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+		response, err = responseFetcher(paginationKey)
+		if err != nil {
+			return nil, err
+		}
+		responses = append(responses, response)
+	}
+
+	return responses, nil
+}
```
Comment (anukul) [inline cosmos/lib/pagination.go:37]: We can add another function argument here to reduce the boilerplate of parsing results from responses (e.g. `response.GetSupply`), but it seems to reduce readability for very little gain

```go
// PaginatedRequest returns the aggregate result of a paginated cosmos request
func PaginatedRequest[T PageResponse, U interface{}](
	responseFetcher func(paginationKey []byte) (T, error),
	resultGetter func(response T) []U,
) ([]U, error) {
	var result []U
	var response T
	var err error

	for paginationKey, i := []byte{}, 0; len(paginationKey) > 0 || i == 0; paginationKey, i = getNextPaginationKey(response), i+1 {
		response, err = responseFetcher(paginationKey)
		if err != nil {
			return nil, err
		}
		result = append(result, resultGetter(response)...)
	}

	return result, nil
}
```

```go
supply, err := cosmlib.PaginatedRequest[*banktypes.QueryTotalSupplyResponse, types.Coin](
	func(paginationKey []byte) (*banktypes.QueryTotalSupplyResponse, error) {
		return c.querier.TotalSupply(ctx, &banktypes.QueryTotalSupplyRequest{
			Pagination: &query.PageRequest{Key: paginationKey},
		})
	},
	func(response *banktypes.QueryTotalSupplyResponse) []types.Coin {
		return response.GetSupply()
	},
)
```

**Comment by itsdevbear** [comment]
looks great! But won't we need to modify the solidity interface to pass in a page number?

**Comment by anukul** [comment]
oh, i was just iterating over all pages. didn't realise this was exposed to the caller, will update.

that should also alleviate the gas concern if there are too many pages.

**Comment by itsdevbear** [comment]
ye exactly, the biggest issue with just yeeting the whole thing, is eventually it'll just consume more gas than what fits in a block and then its gg

**Comment by anukul** [comment]
@itsdevbear updated, only added to `Bank.getAllBalances` to start. can do for other methods once this is good to go :)

File: contracts/src/cosmos/CosmosTypes.sol [hunk 4]
```diff
@@ -36,11 +36,26 @@ library Cosmos {
         uint256 amount;
         string denom;
     }
+
+    struct PageRequest {
+        string key;
```
Comment (anukul) [inline contracts/src/cosmos/CosmosTypes.sol:41]: Open to changing page key to `bytes` 🤔

**Comment by itsdevbear** [comment]
@anukul shoot me a msg on TG.

**Comment by itsdevbear** [comment]
fire af

**Comment by itsdevbear** [comment]
@garrettparris this should be helpful for you

File: cosmos/testing/e2e/precompile/staking/staking_test.go [hunk 5]
```diff
@@ -79,7 +79,7 @@ var _ = Describe("Staking", func() {
 	})
 
 	It("should call functions on the precompile directly", func() {
```
Comment (anukul) [inline cosmos/testing/e2e/precompile/staking/staking_test.go:81]: These tests need to be revised after the e2e localnet node config is updated to have multiple validators

**Comment by calbera** [comment]
@anukul thank you for beginning this work! We used it as inspiration and implemented it for staking and governance precompiles. it would be much easier to work off of main and implement the pagination just for the bank precompile from scratch; sorry about the inconvenience :)

**Pull request closed without merging**